globals/matchmaking_utils.go

```diff
@@ -131,21 +128,171 @@ func RemoveConnectionFromAllSessions(connection *nex.PRUDPConnection) {
 			messagePacket.SetType(constants.DataPacket)
 			messagePacket.AddFlag(constants.PacketFlagNeedsAck)
 			messagePacket.AddFlag(constants.PacketFlagReliable)
-			messagePacket.SetSourceVirtualPortStreamType(connection.StreamType)
+			messagePacket.SetSourceVirtualPortStreamType(target.StreamType)
 			messagePacket.SetSourceVirtualPortStreamID(endpoint.StreamID)
-			messagePacket.SetDestinationVirtualPortStreamType(connection.StreamType)
-			messagePacket.SetDestinationVirtualPortStreamID(connection.StreamID)
+			messagePacket.SetDestinationVirtualPortStreamType(target.StreamType)
+			messagePacket.SetDestinationVirtualPortStreamID(target.StreamID)
 			messagePacket.SetPayload(rmcRequestBytes)
 
 			server.Send(messagePacket)
+
+			return false
+		})
+	}
+	
+	if SessionManagementDebugLog {
+		globals.Logger.Infof("GID %d: Deleted", gathering)
+	}
+	delete(sessions, gathering)
+}
+
+func RemoveSession(connection *nex.PRUDPConnection, gathering uint32) {
+	sessionsMutex.Lock()
+	defer sessionsMutex.Unlock()
+
+	removeSessionImpl(connection, gathering)
+}
+
+// RemoveConnectionIDFromSession removes a PRUDP connection from the session
+func removeConnectionIDFromSessionImpl(connection *nex.PRUDPConnection, gathering uint32, gracefully bool) {
+	session, ok := sessions[gathering]
+	if !ok {
+		return
+	}
+	session.ConnectionIDs.DeleteAll(connection.ID)
+
+	ownerPID := session.GameMatchmakeSession.Gathering.OwnerPID
+	lenParticipants := session.ConnectionIDs.Size()
+
+	if SessionManagementDebugLog {
+		var grace string
+		if gracefully {
+			grace = "gracefully"
+		} else {
+			grace = "ungracefully"
 		}
+		globals.Logger.Infof("GID %d: Removed PID %d %s", gathering, connection.PID().LegacyValue(), grace)
+	}
 
-		gid = FindConnectionSession(connection.ID)
+	// If there are no more participants remove the session
+	if lenParticipants == 0 {
+		removeSessionImpl(connection, gathering)
+		return
+	}
+
+	// If the owner is the one being removed...
+	if ownerPID.Equals(connection.PID()) {
+		// * This flag tells the server to change the matchmake session owner if they disconnect
+		// * If the flag is not set, delete the session
+		// * More info: https://nintendo-wiki.pretendo.network/docs/nex/protocols/match-making/types#flags
+		if session.GameMatchmakeSession.Gathering.Flags.PAND(match_making.GatheringFlags.DisconnectChangeOwner) == 0 {
+			removeSessionImpl(connection, gathering)
+			return
+		} else {
+			// TODO: Figure out if this is the correct way to do it!
+			changeSessionOwnerImpl(connection, gathering, true)
+		}
+	}
+	if !gracefully {
+		// Only send the disconnected notification if the connection disonnected ungracefully
+		endpoint := connection.Endpoint().(*nex.PRUDPEndPoint)
+		server := endpoint.Server
+
+		category := notifications.NotificationCategories.Participation
+		subtype := notifications.NotificationSubTypes.Participation.Disconnected
```

Comment:
We could inline the "graceful" notification on here, so that we don't have to handle it on `EndParticipation` and avoid race conditions

```suggestion
	var subtype uint32
	if graceful {
		subtype = notifications.NotificationSubTypes.Participation.Ended
	} else {
		subtype = notifications.NotificationSubTypes.Participation.Disconnected
	}
```